### ISSUE TITLE
Support reloading when global options change 

### ISSUE THREAD

**Issue by ktff** [timberio/vector#1815]
### Context
Vector supports being reloaded while running, but not if any global option has been changed. 

### Change
We should support reloading even if some global option has been changed.

### Implementation
This could be as simple as considering everything has changed when any global option has been changed. But complications could arise when global option `data_dir` is changed regards to https://vector.dev/docs/administration/process-management/#graceful-pipeline-transitioning

### Motivation
Removing this exception would streamline our reload story, which would in turn simplify our documentation and enhance UX.

**Comment by binarylogic** [comment]
Thanks for opening this. I agree! I'm curious how often users are changing global variables like this. I'd like to hear more actual use cases before work begins.

**Comment by ktff** [comment]
I agree. Especially because, with the introduction of `log schema`, the implementation has become more complicated.

**Comment by tshepang** [comment]
It probably happens more for early users, as they are exploring the tool. I didn't find the requirement annoying though... it is indeed rare to change global config. I consider it a nice-to-have.

**Comment by jszwedko** [comment]
I discussed this today with @lukesteensen today and he made the good observation that maybe we should split out options, like `data_dir`, that are very difficult to handle reload for to CLI flags or, potentially, a separate file, thus making it clearer which options are reloadable and which are not. Generally reload is most useful for updates to the topology itself.

Related:
- https://github.com/vectordotdev/vector/pull/17784

**Comment by suikammd** [comment]
Maybe we can add a force-reload flag when global options change? I can help make it if you want this feature. @jszwedko

**Comment by pront** [comment]
Hi @suikammd, in regards to https://github.com/vectordotdev/vector/pull/23500#issuecomment-3144464081, suppose we set `force_reload` to true, what do you expect to see? There's still the problem of gracefully handling a `data_dir` change. I am thinking it's safer to due a graceful shutdown and then start again.

**Comment by suikammd** [comment]
Hi @pront ,

Thanks for the response! You're absolutely right that changing data_dir is risky and should ideally trigger a restart.

That said, from a platform integration perspective (especially in containerized environments), it would still be helpful to have a structured signal when a config reload is rejected due to GlobalOptions changes — even if we don’t allow hot reload.

For example, emitting a metric like vector_config_reload_rejected{reason="global_options"} would allow us to observe this state and react accordingly (e.g. trigger a rolling restart).

Right now, the only way to detect such a reload failure is by parsing logs, which is brittle and makes automation more difficult. A dedicated metric or internal event would make platform-level coordination much cleaner.

Would the team be open to exposing such a metric or signal, separately from implementing force_reload?

Thanks again!

**Comment by pront** [comment]
> For example, emitting a metric like vector_config_reload_rejected{reason="global_options"} would allow us to observe this state and react accordingly (e.g. trigger a rolling restart).
> 
> Right now, the only way to detect such a reload failure is by parsing logs, which is brittle and makes automation more difficult. A dedicated metric or internal event would make platform-level coordination much cleaner.

This sounds like a good idea, yes.

**Comment by suikammd** [comment]
> > For example, emitting a metric like vector_config_reload_rejected{reason="global_options"} would allow us to observe this state and react accordingly (e.g. trigger a rolling restart).
> > Right now, the only way to detect such a reload failure is by parsing logs, which is brittle and makes automation more difficult. A dedicated metric or internal event would make platform-level coordination much cleaner.
> 
> This sounds like a good idea, yes.

Hi @pront , thanks again for the feedback! I've added the internal event and metric as discussed — vector_config_reload_rejected{reason="global_options"}  will now be emitted when reload is rejected due to GlobalOptions changes. You may check the pr #23500.

Feel free to take a look and let me know if you'd suggest any adjustments!
